src/Raven.Server/Documents/AttachmentsStorage.cs

```diff
@@ -1359,8 +1363,37 @@ public static IEnumerable<BlittableJsonReaderObject> GetAttachmentsFromDocumentM
         [StorageIndexEntryKeyGenerator]
         internal static ByteStringContext.Scope GenerateBucketAndEtagIndexKeyForAttachments(ByteStringContext context, ref TableValueReader tvr, out Slice slice)
         {
-            return ExtractIdFromKeyAndGenerateBucketAndEtagIndexKey(context, (int)AttachmentsTable.LowerDocumentIdAndLowerNameAndTypeAndHashAndContentType,
+            return ShardedDocumentsStorage.ExtractIdFromKeyAndGenerateBucketAndEtagIndexKey(context, (int)AttachmentsTable.LowerDocumentIdAndLowerNameAndTypeAndHashAndContentType,
                 (int)AttachmentsTable.Etag, ref tvr, out slice);
         }
+
+        private void UpdateBucketStatsOnPutOrDeleteStream(DocumentsOperationContext context, Slice attachmentKey, long sizeChange)
+        {
+            if (_documentDatabase is not ShardedDocumentDatabase)
+                return;
+
+            using (GetBucketFromAttachmentKey(context, attachmentKey, out var bucketSlice))
+            {
+                ShardedDocumentsStorage.UpdateBucketStats(context.Transaction.InnerTransaction, bucketSlice, oldSize : 0, newSize : sizeChange);
+            }
+        }
+
+        private static ByteStringContext.Scope GetBucketFromAttachmentKey(DocumentsOperationContext context, Slice attachmentKey, out Slice bucketSlice)
+        {
+            int sizeOfDocId = 0;
+            for (; sizeOfDocId < attachmentKey.Size; sizeOfDocId++)
+            {
+                if (attachmentKey[sizeOfDocId] == SpecialChars.RecordSeparator)
+                    break;
+            }
+
+            var bucket = ShardHelper.GetBucket(attachmentKey.Content.Ptr, sizeOfDocId);
+            var scope = context.Allocator.Allocate(sizeof(int), out var buffer);
+            *(int*)buffer.Ptr = Bits.SwapBytes(bucket);
+
+            bucketSlice = new Slice(buffer);
+
+            return scope;
+        }
```

Comment:
let's move this and all sharded related functions into a new partial class `AttachmentsStorage.Sharding`

Reply:
done